sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,357 @@
+      ByteString bytes = entry.getValue().getOutput().toByteString();
+      if (bytes.isEmpty()) {
+        continue;
+      }
+      bufferedElements
+          .addTimersBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(entry.getKey().getKey())
+          .setTimerFamilyId(entry.getKey().getValue())
+          .setTimers(bytes);
+      entry.getValue().reset();
+      perEndpointElementCount.compute(
+          LogicalEndpoint.timer(
+              processBundleRequestIdSupplier.get(),
+              entry.getKey().getKey(),
+              entry.getKey().getValue()),
+          (e, count) -> count == null ? 1 : count + 1);
+    }
+    totalByteCounter.reset();
+    return bufferedElements;
+  }
+
+  private void periodicFlush() {
+    try {
+      synchronized (flushLock) {
+        flush();
+      }
+    } catch (Throwable t) {
+      throw new RuntimeException(t);
+    }
+  }
+
+  /** Check if the flush thread failed with an exception. */
+  private void checkFlushThreadException() throws IOException {
+    if (timeLimit > 0 && flushFuture.isDone()) {
+      try {
+        flushFuture.get();
+        throw new IOException("Periodic flushing thread finished unexpectedly.");
+      } catch (ExecutionException ee) {
+        unwrapExecutionException(ee);
+      } catch (CancellationException ce) {
+        throw new IOException(ce);
+      } catch (InterruptedException ie) {
+        Thread.currentThread().interrupt();
+        throw new IOException(ie);
+      }
+    }
+  }
+
+  private void unwrapExecutionException(ExecutionException ee) throws IOException {
+    // the cause is always RuntimeException
+    RuntimeException re = (RuntimeException) ee.getCause();
+    if (re.getCause() instanceof IOException) {
+      throw (IOException) re.getCause();
+    } else {
+      throw new IOException(re.getCause());
+    }
+  }
+
+  private static int getSizeLimit(PipelineOptions options) {
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(DATA_BUFFER_SIZE_LIMIT)) {
+        return Integer.parseInt(experiment.substring(DATA_BUFFER_SIZE_LIMIT.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT_BYTES;
+  }
+
+  private static long getTimeLimit(PipelineOptions options) {
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(DATA_BUFFER_TIME_LIMIT_MS)) {
+        return Long.parseLong(experiment.substring(DATA_BUFFER_TIME_LIMIT_MS.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT_TIME_MS;
+  }
+
+  private static class Receiver<T> implements FnDataReceiver<T> {
```

Comment:
If you make this class non static you can put the flush code into `accept` since it will have access to the methods and member variables of the parent.
```suggestion
  private class Receiver<T> implements FnDataReceiver<T> {
```

This also means we don't need to store a reference to the `totalByteCounter` since we can access the enclosing class.